crates/pgt_typecheck/src/typed_identifier.rs

```diff
@@ -0,0 +1,340 @@
+    /// The path of the parameter, usually the name of the function.
+    /// This is because `fn_name.arg_name` is a valid reference within a SQL function.
+    pub path: String,
+    /// The name of the argument
+    pub name: Option<String>,
+    /// The type of the argument with schema and name
+    pub type_: IdentifierType,
+}
+
+#[derive(Debug, Clone)]
+pub struct IdentifierType {
+    pub schema: Option<String>,
+    pub name: String,
+    pub is_array: bool,
+}
+
+/// Applies the identifiers to the SQL string by replacing them with their default values.
+pub fn apply_identifiers<'a>(
+    identifiers: Vec<TypedIdentifier>,
+    schema_cache: &'a pgt_schema_cache::SchemaCache,
+    cst: &'a tree_sitter::Tree,
+    sql: &'a str,
+) -> (String, bool) {
+    let mut executor = TreeSitterQueriesExecutor::new(cst.root_node(), sql);
+
+    executor.add_query_results::<ParameterMatch>();
+
+    // Collect all replacements first to avoid modifying the string while iterating
+    let replacements: Vec<_> = executor
+        .get_iter(None)
+        .filter_map(|q| {
+            let m: &ParameterMatch = q.try_into().ok()?;
+            let path = m.get_path(sql);
+            let parts: Vec<_> = path.split('.').collect();
+
+            // Find the matching identifier and its position in the path
+            let (identifier, position) = find_matching_identifier(&parts, &identifiers)?;
+
+            // Resolve the type based on whether we're accessing a field of a composite type
+            let type_ = resolve_type(identifier, position, &parts, schema_cache)?;
+
+            Some((m.get_byte_range(), type_, identifier.type_.is_array))
+        })
+        .collect();
+
+    let mut result = sql.to_string();
+
+    let mut valid_positions = true;
+
+    // Apply replacements in reverse order to maintain correct byte offsets
+    for (range, type_, is_array) in replacements.into_iter().rev() {
+        let default_value = get_formatted_default_value(type_, is_array);
+
+        // if the default_value is shorter than "range", fill it up with spaces
+        let default_value = if default_value.len() < range.end - range.start {
+            format!("{:<width$}", default_value, width = range.end - range.start)
+        } else {
+            default_value
+        };
+
+        // check if we can maintain a valid position
+        if default_value.len() > range.end - range.start {
+            valid_positions = false;
+        }
+
+        result.replace_range(range, &default_value);
+    }
+
+    (result, valid_positions)
+}
+
+/// Format the default value based on the type and whether it's an array
+fn get_formatted_default_value(pg_type: &PostgresType, is_array: bool) -> String {
+    // Get the base default value for this type
+    let default = resolve_default_value(pg_type);
+
+    let default = if default.len() > "NULL".len() {
+        // If the default value is longer than "NULL", use "NULL" instead
+        "NULL".to_string()
```

Comment:
why are we doing this?

Reply:
because we need to throw away positions when the replacement is longer than its original, and `NULL` will be shorter than a few default values (e.g. timestamps or custom enums)